roles/oVirt.image-template/library/ovirt_templates.py

```diff
@@ -0,0 +1,619 @@
+        )
+    return domainMappings
+
+
+def _get_cluster_mappings(module):
+    clusterMappings = list()
+
+    for clusterMapping in module.params['cluster_mappings']:
+        clusterMappings.append(
+            otypes.RegistrationClusterMapping(
+                from_=otypes.Cluster(
+                    name=clusterMapping['source_name'],
+                ),
+                to=otypes.Cluster(
+                    name=clusterMapping['dest_name'],
+                ),
+            )
+        )
+    return clusterMappings
+
+
+def _get_vnic_profile_mappings(module):
+    vnicProfileMappings = list()
+
+    for vnicProfileMapping in module.params['vnic_profile_mappings']:
+        vnicProfileMappings.append(
+            otypes.VnicProfileMapping(
+                source_network_name=vnicProfileMapping['source_network_name'],
+                source_network_profile_name=vnicProfileMapping['source_profile_name'],
+                target_vnic_profile=otypes.VnicProfile(
+                    id=vnicProfileMapping['target_profile_id'],
+                ) if vnicProfileMapping['target_profile_id'] else None,
+            )
+        )
+
+    return vnicProfileMappings
+
+
+def searchable_attributes(module):
+    """
+    Return all searchable template attributes passed to module.
+    """
+    attributes = {
+        'name': module.params.get('name'),
+        'cluster': module.params.get('cluster'),
+    }
+    return {k: v for k, v in attributes.items() if v is not None}
+
+
+def main():
+    argument_spec = ovirt_full_argument_spec(
+        state=dict(
+            choices=['present', 'absent', 'exported', 'imported', 'registered'],
+            default='present',
+        ),
+        id=dict(default=None),
+        name=dict(default=None),
+        vm=dict(default=None),
+        description=dict(default=None),
+        cluster=dict(default=None),
+        allow_partial_import=dict(default=None, type='bool'),
+        cpu_profile=dict(default=None),
+        disks=dict(default=[], type='list'),
+        clone_permissions=dict(type='bool'),
+        export_domain=dict(default=None),
+        storage_domain=dict(default=None),
+        exclusive=dict(type='bool'),
+        image_provider=dict(default=None),
+        image_disk=dict(default=None, aliases=['glance_image_disk_name']),
+        template_image_disk_name=dict(default=None),
+        seal=dict(type='bool'),
+        vnic_profile_mappings=dict(default=[], type='list'),
+        cluster_mappings=dict(default=[], type='list'),
+        role_mappings=dict(default=[], type='list'),
+        domain_mappings=dict(default=[], type='list'),
+        operating_system=dict(type='str'),
+        memory=dict(type='str'),
+        memory_guaranteed=dict(type='str'),
+        memory_max=dict(type='str'),
+    )
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_one_of=[['id', 'name']],
+    )
+    check_sdk(module)
+
+    try:
```

Comment:
Having one large try: except: block for the entire module will make debugging quite burdensome.

Reply:
We have traceback, so it's OK.

Reply:
It's confusing for most users to try to digest a raw traceback.  It's much better user experience to leave something helpful like "unable to connect to url."  While you or I could probably figure out just about any error message this module is likely to throw at us, the end user might not.

I'm not going to block on this because I don't know how big of a refactor it is for you, but it needs to be cleaned up within the next few sprints.

Reply:
Well in case of any error we throw meanigful error message right from the Python SDK, for example in case of connection failed:

```json
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_dMmrLy/ansible_module_ovirt_auth.py", line 289, in main
    token = connection.authenticate()
  File "/usr/lib64/python2.7/site-packages/ovirtsdk4/__init__.py", line 384, in authenticate
    self.__parse_error(e)
  File "/usr/lib64/python2.7/site-packages/ovirtsdk4/__init__.py", line 932, in __parse_error
    six.reraise(clazz, clazz(error_msg), sys.exc_info()[2])
  File "/usr/lib64/python2.7/site-packages/ovirtsdk4/__init__.py", line 381, in authenticate
    self._sso_token = self._get_access_token()
  File "/usr/lib64/python2.7/site-packages/ovirtsdk4/__init__.py", line 617, in _get_access_token
    sso_response = self._get_sso_response(self._sso_url, post_data)
  File "/usr/lib64/python2.7/site-packages/ovirtsdk4/__init__.py", line 694, in _get_sso_response
    curl.perform()
ConnectionError: Error while sending HTTP request: (7, 'Failed to connect to fqdn port 8443: Connection refused')

fatal: [127.0.0.1]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "ca_file": null, 
            "compress": true, 
            "headers": null, 
            "hostname": null, 
            "insecure": true, 
            "kerberos": false, 
            "ovirt_auth": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "state": "present", 
            "timeout": 0, 
            "token": null, 
            "url": "https://fqdn/ovirt-engine/api", 
            "username": "admin@internal"
        }
    }, 
    "msg": "Error while sending HTTP request: (7, 'Failed to connect to fqdn port 8443: Connection refused')"
}
```

The same looks like any other errors, they are returned from oVirt API backend and Ansible just show them together with traceback.